src/script/auth/component/EntropyCanvas.tsx

```diff
@@ -19,89 +19,90 @@
+  const [lastPoint, setLastPoint] = useState<EntropyFrame | null>(null);
 
   const {clearInterval, startInterval, pauseInterval} = usePausableInterval(() => {
-    // This prevents automatic batching of state updates from react 18. https://github.com/reactwg/react-18/discussions/21
-    flushSync(() => {
-      setPercent(percent => percent + 1);
-    });
+    setPercent(Math.floor(100 * Math.min(entropy.entropyBits / min_entropy_bits, entropy.length / min_frames)));
     onProgress(entropy, percent, false);
-  }, 300);
+  }, 100);
 
   useEffect(() => {
-    if (frames <= 300 && percent > 95) {
-      setPercent(95);
-      pauseInterval();
-    }
-    if (frames > 300) {
-      startInterval();
-    }
-    if (frames >= 100) {
+    if (percent >= 100) {
       clearInterval();
+      onProgress(entropy, percent, false);
     }
-  }, [percent, percent]);
+  }, [clearInterval, percent]);
 
   const onMouseLeave = () => {
     onProgress(entropy, percent, true);
     pauseInterval();
-
-    setEntropy([...entropy, null]);
+    // add null value to prevent render artifacts when the mouse leaves on one side and enters at a different position
+    setPreviousPoint(lastPoint);
+    setLastPoint(null);
   };
 
-  const draw = (ctx: CanvasRenderingContext2D, entropy: [number, number][]) => {
-    if (entropy.length > 2) {
-      const previousPoint = entropy[entropy.length - 2];
-      const lastPoint = entropy[entropy.length - 1];
-      ctx.beginPath();
-      if (!previousPoint) {
-        ctx.moveTo(...lastPoint);
-      } else {
-        ctx.moveTo(...previousPoint);
-      }
-      ctx.lineJoin = 'round';
-      ctx.lineCap = 'round';
-      ctx.lineWidth = 2;
-      ctx.strokeStyle = 'blue';
-      if (!lastPoint) {
-        ctx.lineTo(...previousPoint);
-      } else {
-        ctx.lineTo(...lastPoint);
-      }
-      ctx.stroke();
+  const draw = (ctx: CanvasRenderingContext2D) => {
+    if (!previousPoint || !lastPoint) {
+      return;
     }
+    ctx.beginPath();
+    ctx.moveTo(previousPoint.x, previousPoint.y);
+    ctx.lineJoin = 'round';
+    ctx.lineCap = 'round';
+    ctx.lineWidth = 2;
+    ctx.strokeStyle = 'blue';
+    ctx.lineTo(lastPoint.x, lastPoint.y);
+    ctx.stroke();
   };
 
   useEffect(() => {
     const canvas = canvasRef.current;
+    if (!canvas) {
+      return;
+    }
     const context = canvas.getContext('2d');
-    draw(context, entropy);
-  }, [entropy]);
+    if (!context) {
+      return;
+    }
+    draw(context);
+  }, [lastPoint]);
 
   const onMouseMove = (event: MouseEvent<HTMLCanvasElement>) => {
     startInterval();
-    const newEntropy: [number, number] = [
-      event.pageX - event.currentTarget?.getBoundingClientRect()?.x,
-      event.pageY - event.currentTarget?.getBoundingClientRect()?.y,
-    ];
-    setEntropy(entropy => [...entropy, newEntropy]);
+    const boundingRect = event.currentTarget?.getBoundingClientRect();
+    const drawPoint: EntropyFrame = {
+      t: 0,
```

Comment:
is there a reason for t to be included in here besides its inclusion in the type? i suggest to make it an optional parameter in the type or use a different type perhaps.

Reply:
it was mainly because of the type, made it optional now